### PR TITLE
crypto: add SSE-KMS HTTP header detection

### PR DESCRIPTION
**Pull request by aead**
## Description
This commit adds support for detecting SSE-KMS headers.
The server should be able to detect SSE-KMS headers to
at least fail such S3 requests with not implemented.

## Motivation and Context
SSE-KMS

## How Has This Been Tested?
unit tests: `go test -v github.com/minio/minio/cmd/crypto`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by donatello — APPROVED** [review]

**Comment by aead** [comment]
Ping @vadmeste

File: cmd/crypto/header.go [hunk 1]
```diff
@@ -79,6 +96,27 @@ func (s3) ParseHTTP(h http.Header) (err error) {
...
+// S3KMS represents AWS SSE-KMS. It provides functionality to
+// handle SSE-KMS requests.
+var S3KMS = s3KMS{}
+
+type s3KMS struct{}
+
+// IsRequested returns true if the HTTP headers indicates that
+// the S3 client requests SSE-KMS.
+func (s3KMS) IsRequested(h http.Header) bool {
+	if _, ok := h[SSEKmsID]; ok {
```
Comment (kannappanr) [inline cmd/crypto/header.go:108]: In the case of `KMS` I thought that `X-Amz-Server-Side-Encryption-Aws-Kms-Key-Id` and `X-Amz-Server-Side-Encryption` are required and `X-Amz-Server-Side-Encryption`'s value will always be equal to `aws:kms`. `X-Amz-Server-Side-Encryption-Context` is the optional value, so we would not have to check for that. https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html

Comment (aead) [reply cmd/crypto/header.go:108]: @kannappanr Correct. According to the documentation of this function it *indicates whether a S3 client requests SSE-KMS*. If a client specify `X-Amz-Server-Side-Encryption-Context` it indicates that SSE-KMS is requested - whether the (SSE-KMS) request is well-formed is not the responsibility of this function.

We have to detect a SSE request - even if it is malformed to respond with an appropriate error.

Comment (kannappanr) [reply cmd/crypto/header.go:108]: ok. Makes sense now.

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**